src/fides/api/schemas/locations.py

```diff
@@ -0,0 +1,168 @@
+    location_ids_mapped_to_location: Dict[str, Location] = load_locations()
+
+    location_groups_to_location_set: Dict[str, Set[str]] = defaultdict(set)
+
+    for location in location_ids_mapped_to_location.values():
+        if not location.belongs_to:
+            continue
+
+        for belongs_to in location.belongs_to:
+            location_groups_to_location_set[belongs_to].add(location.id)
+
+    return location_groups_to_location_set
+
+
+@lru_cache(maxsize=1)
+def load_regulations() -> Dict[str, LocationRegulationBase]:
+    """Loads regulations dict based on yml file on disk"""
+    with open(load_file([LOCATIONS_YAML_FILE_PATH]), "r", encoding="utf-8") as file:
+        _regulations = yaml.safe_load(file).get("regulations", [])
+        regulation_dict = {}
+        for regulation in _regulations:
+            regulation_dict[regulation["id"]] = LocationRegulationBase.parse_obj(
+                regulation
+            )
+        return regulation_dict
+
+
+locations_by_id: Dict[
+    str, Location
+] = load_locations()  # should only be accessed for read-only access
+location_groups: Dict[
+    str, LocationGroup
+] = load_location_groups()  # should only be accessed for read-only access
+location_group_to_location: Dict[
+    str, Set[str]
+] = load_location_group_to_location()  # should only be accessed for read-only access
+
+
+def _load_privacy_notice_regions() -> Dict[str, Union[Location, LocationGroup]]:
+    """Merge Location dictionary with LocationGroup Dictionary"""
+    return {**locations_by_id, **location_groups}
+
+
+supported_locations_by_id: Dict[
+    str, Union[Location, LocationGroup]
+] = _load_privacy_notice_regions()  # should only be accessed for read-only access
+
+
+# dynamically create an enum based on definitions loaded from YAML
+PrivacyNoticeRegion: Enum = Enum(  # type: ignore[misc]
+    "PrivacyNoticeRegion",
+    {location.id: location.id for location in supported_locations_by_id.values()},
+)
+
+
+def filter_regions_by_location(
+    db: Session,
+    regions: List[PrivacyNoticeRegion],
+) -> List[PrivacyNoticeRegion]:
+    """Filter a list of PrivacyNoticeRegion to only ones that match at the configured Location or LocationGroup level.
+
+    Locations are stored at the Location level while regions can be saved at the Location level
+    or the Location group level.
+    """
+
+    locations: Set[str] = LocationRegulationSelections.get_selected_locations(db)
+
+    if not locations:
+        return regions  # type: ignore[attr-defined]
+
+    filtered_regions: List[PrivacyNoticeRegion] = []
+
+    for region in regions:
+        # Explode any Location Groups to Locations so they can be compared against apples to apples
+        # with locations
+        expanded_regions: Set[str] = location_group_to_location.get(
+            region.value, {region.value}
+        )
+        if expanded_regions.issubset(locations):
+            filtered_regions.append(region)
```

Comment:
ah yes, this is a good edge case - nicely handled! the logic/behavior here seems correct. i think the only tweak i'd consider is whether we want to more properly store the "selected location groups" as a general-purpose method on the LocationRegulationSelections model, so we don't have this logic buried within this helper function.

basically, would it make sense for us to be able to call LocationRegulationSelections.get_selected_location_groups(db) and then compare against that here a bit more directly? that just feels a bit more transparent to me.

the LocationRegulationSelections.get_selected_location_groups method would, underneath the hood, probably do something similar to what's done here, or perhaps we'd take some other measures when actually making any updates to selected locations to ensure that the selected location groups are updated accordingly at that time, i.e. to prevent repeatedly "recalculating" the selected groups, for performance reasons

another middle-ground would be to just pull some of this logic here into its utility function on the LocationRegulationSelections module. i think that's the main thing i'm after - to pull this logic into a function/method that's a bit more visible and easy for other use cases to access as needed.

let me know if that makes sense!

Reply:
OK progress so far! I've updated to store both locations and location groups from the API endpoints like we talked about.  PATCH locations endpoint looks at locations passed in and combines them into location groups where applicable and saves them into a new column.  Also updated regulation changes to locations can also propagate back to location groups.

That makes this existing funciton a lot simpler, to compare PrivacyNoticeRegions to a union of saved Locations and LocationGroups.

I've left this existing method where it is for circular import regions although it can be moved to a different location if needed.